Breakages on newer versions of Clang

I work on the ChromeOS toolchain team, and while upgrading LLVM from `a58d0af` to `db1978b`, we've noticed that cvise no longer builds due to the following issues.

```
[15/87] Building CXX object clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o
FAILED: clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o 
/usr/bin/x86_64-pc-linux-gnu-clang++ -DHAVE_CONFIG_H -I. -I../clang_delta  -DNDEBUG -std=c++14 -fno-rtti -fno-strict-aliasing -Wall -Wextra -Wno-unused-parameter -Werror -Wno-error=maybe-uninitialized -fvisibility-inlines-hidden   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o -MF clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o.d -o clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o -c ../clang_delta/TransformationManager.cpp
../clang_delta/TransformationManager.cpp:157:16: error: no member named 'setLangDefaults' in 'clang::CompilerInvocation'
    Invocation.setLangDefaults(ClangInstance->getLangOpts(), InputKind(Language::C), T, includes);
    ~~~~~~~~~~ ^
../clang_delta/TransformationManager.cpp:163:16: error: no member named 'setLangDefaults' in 'clang::CompilerInvocation'
    Invocation.setLangDefaults(ClangInstance->getLangOpts(), InputKind(Language::CXX), T, includes, LSTD);
    ~~~~~~~~~~ ^
../clang_delta/TransformationManager.cpp:194:16: error: no member named 'setLangDefaults' in 'clang::CompilerInvocation'
    Invocation.setLangDefaults(ClangInstance->getLangOpts(),
    ~~~~~~~~~~ ^
3 errors generated.
[36/87] Building CXX object clang_delta/CMakeFiles/clang_delta.dir/ExpressionDetector.cpp.o
FAILED: clang_delta/CMakeFiles/clang_delta.dir/ExpressionDetector.cpp.o 
/usr/bin/x86_64-pc-linux-gnu-clang++ -DHAVE_CONFIG_H -I. -I../clang_delta  -DNDEBUG -std=c++14 -fno-rtti -fno-strict-aliasing -Wall -Wextra -Wno-unused-parameter -Werror -Wno-error=maybe-uninitialized -fvisibility-inlines-hidden   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT clang_delta/CMakeFiles/clang_delta.dir/ExpressionDetector.cpp.o -MF clang_delta/CMakeFiles/clang_delta.dir/ExpressionDetector.cpp.o.d -o clang_delta/CMakeFiles/clang_delta.dir/ExpressionDetector.cpp.o -c ../clang_delta/ExpressionDetector.cpp
../clang_delta/ExpressionDetector.cpp:62:16: error: 'InclusionDirective' marked 'override' but does not override any member functions
  virtual void InclusionDirective(SourceLocation HashLoc,
               ^
../clang_delta/ExpressionDetector.cpp:62:16: error: '(anonymous namespace)::IncludesPPCallbacks::InclusionDirective' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
/usr/include/clang/Lex/PPCallbacks.h:105:16: note: hidden overloaded virtual function 'clang::PPCallbacks::InclusionDirective' declared here: type mismatch at 6th parameter ('Optional<clang::FileEntryRef>' vs 'const clang::FileEntry *')
  virtual void InclusionDirective(SourceLocation HashLoc,
               ^
2 errors generated.
[86/87] Building CXX object clang_delta/CMakeFiles/clang_delta.dir/RemoveUnusedFunction.cpp.o
ninja: build stopped: subcommand failed.
```

--- a/clang_delta/ExpressionDetector.cpp
+++ b/clang_delta/ExpressionDetector.cpp
@@ -16,6 +16,9 @@
 
 #include "clang/AST/RecursiveASTVisitor.h"
 #include "clang/AST/ASTContext.h"
+#if LLVM_VERSION_MAJOR >= 15
+#include "clang/Basic/FileEntry.h"
+#endif
 #include "clang/Basic/SourceManager.h"
 #include "clang/Lex/Preprocessor.h"
 
@@ -62,7 +65,12 @@ public:
   virtual void InclusionDirective(SourceLocation HashLoc,
                           const Token &IncludeTok,
                           StringRef FileName, bool IsAngled,
-                          CharSourceRange FilenameRange, const FileEntry *File,
+                          CharSourceRange FilenameRange,
+#if LLVM_VERSION_MAJOR < 15
+                          const FileEntry *File,
+#else
+                          Optional<FileEntryRef> File,
+#endif
                           StringRef SearchPath, StringRef RelativePath,
                           const Module *Imported,
                           SrcMgr::CharacteristicKind FileType) override;
@@ -80,9 +88,13 @@ private:
 void IncludesPPCallbacks::InclusionDirective(SourceLocation HashLoc,
                                              const Token &/*IncludeTok*/,
                                              StringRef FileName,
-                                            bool /*IsAngled*/,
+                                             bool /*IsAngled*/,
                                              CharSourceRange /*FilenameRange*/,
+#if LLVM_VERSION_MAJOR < 15
                                              const FileEntry * /*File*/,
+#else
+                                             Optional<FileEntryRef> /*File*/,
+#endif
                                              StringRef /*SearchPath*/,
                                              StringRef /*RelativePath*/,
                                              const Module * /*Imported*/,

--- a/clang_delta/TransformationManager.cpp
+++ b/clang_delta/TransformationManager.cpp
@@ -20,6 +20,10 @@
 #include "clang/Basic/Builtins.h"
 #include "clang/Basic/Diagnostic.h"
 #include "clang/Basic/FileManager.h"
+#if LLVM_VERSION_MAJOR >= 15
+#include "clang/Basic/LangOptions.h"
+#include "clang/Basic/LangStandard.h"
+#endif
 #include "clang/Basic/TargetInfo.h"
 #include "clang/Lex/Preprocessor.h"
 #include "clang/Frontend/CompilerInstance.h"
@@ -151,7 +155,7 @@ bool TransformationManager::initializeCompilerInstance(std::string &ErrorMsg)
     Invocation.setLangDefaults(ClangInstance->getLangOpts(), InputKind(Language::CXX), T, PPOpts, LSTD);
   }
   else if(IK.getLanguage() == Language::OpenCL) {
-#else
+#elif LLVM_VERSION_MAJOR < 15
   vector<string> includes;
   if (IK.getLanguage() == Language::C) {
     Invocation.setLangDefaults(ClangInstance->getLangOpts(), InputKind(Language::C), T, includes);
@@ -161,6 +165,18 @@ bool TransformationManager::initializeCompilerInstance(std::string &ErrorMsg)
     // for a function which has a non-declared callee, e.g.,
     // It results an empty AST for the caller.
     Invocation.setLangDefaults(ClangInstance->getLangOpts(), InputKind(Language::CXX), T, includes, LSTD);
+  }
+  else if(IK.getLanguage() == Language::OpenCL) {
+#else
+  vector<string> includes;
+  if (IK.getLanguage() == Language::C) {
+    LangOptions::setLangDefaults(ClangInstance->getLangOpts(), Language::C, T, includes);
+  }
+  else if (IK.getLanguage() == Language::CXX) {
+    // ISSUE: it might cause some problems when building AST
+    // for a function which has a non-declared callee, e.g.,
+    // It results an empty AST for the caller.
+    LangOptions::setLangDefaults(ClangInstance->getLangOpts(), Language::CXX, T, includes, LSTD);
   }
   else if(IK.getLanguage() == Language::OpenCL) {
 #endif
@@ -191,11 +207,18 @@ bool TransformationManager::initializeCompilerInstance(std::string &ErrorMsg)
                                        &Args[0], &Args[0] + Args.size(),
 #endif
                                        ClangInstance->getDiagnostics());
+#if LLVM_VERSION_MAJOR < 15
     Invocation.setLangDefaults(ClangInstance->getLangOpts(),
-#if LLVM_VERSION_MAJOR >= 10
+#else
+    LangOptions::setLangDefaults(ClangInstance->getLangOpts(),
+#endif
+
+#if LLVM_VERSION_MAJOR < 10
+                               InputKind::OpenCL,
+#elif LLVM_VERSION_MAJOR < 15
                                InputKind(Language::OpenCL),
 #else
-                               InputKind::OpenCL,
+                               Language::OpenCL,
 #endif
 
 #if LLVM_VERSION_MAJOR < 12

